<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>

Привет Иван! Меня зовут Марат, и я буду твоим ревьюером. Спешу сообщить что все ключевые этапы в работе выполнены,  с задачей тебе удалось справиться. По поводу обращения - в IT сфере принято общаться на «ты» :) Но, если привычней на «вы», дай знать. Как ревьюера моя задача помочь тебе в развитии, дав хорошие советы. Я внимательно посмотрю твой код, ознакомлюсь с твоими выводами и оставлю комментарии. Где то могу предложить небольшие исправление в коде, но ненавязчиво. Где потребуются уточнения, я оставлю много наводящих вопросов. Они помогут тебя с поиском верного решения.

Все мои комментарии размечены по цветам, для лучшего восприятия. 
    
<div class="alert alert-success">Зеленым цветом и словом «Успех» отмечены особо удачные и элегантные решения, которыми ты можешь гордиться. </div>
        
<div class="alert alert-warning">Желтым и значком словом «Совет», помечены решения у которых есть альтернативные решения, более оптимальные. Ты можешь найти их сразу и доработать проект, или отложить это на потом, для будущих проектах. Проект будет принят и без их доработки. </div>
        
<div class="alert alert-danger"> Красным цветом и значком словом «Ошибка» помечу твои решения, на которые стоит обратить внимание прежде всего. После их доработки проект будет принят. </div>
        
Залог успеха - работа сообща, взаимное уважение и работа в диалоге. Поэтому, помечай свои ответные комментарии на мои реплики заметным цветом или курсивом, так мне будет легче их отслеживать. Пожалуйста, не изменяй и не удаляй мои комментарии. Все это поможет выполнить повторную проверку быстрей.

</div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Вступление в работу очень важно, так человек, который смотрит твой проект (и на работе в том числе) будет сразу введен в курс дела.





</div>

## Откройте и изучите файл

Импорт нужных библиотек для работы с данными и чтение самих данных из файла

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import time
import warnings
warnings.filterwarnings('ignore')
from sklearn.metrics import f1_score

df = pd.read_csv('/datasets/users_behavior.csv')

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Собираем все импорты в верхней части, чтобы легче было ориентироваться и добавлять новые по необходимости. 


<div class="alert alert-warning">

Совет:     
    


- Иван, все-таки стоит импорт библиотек и выгрузку данных разнести в разные ячейки, так стилистически более правильно.


- кстати есть рекомендации PEP-8 при написании кода, в том числе и для импортов. Если интересно можешь почитать [тут](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html), в будущем пригодится )  



<div>

Выяснение основной информации о данных: какой столбец относится к определённому типу данных, что в них содержится.

In [2]:
display(df.info())
display(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


None

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

Проверка данных на пустые значения

In [3]:
df.isnull().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

Изучение файла показало, что в действительности нет необходимости в предобработке данных.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

👍 Данные изучены.


    

<div class="alert alert-warning">



Совет: 


- Можно еще отдельно проверить датасет на сбалансированность классов в таргете. Это помогло бы решить бонусное задание сразу (Только не надо балансировать данные, это тема следующего проекта).




- Можно посмотреть корреляцию признаков. Знаешь что такое мультиколлинеарность, какие два типа проблем возникает, для каких моделей и какие варианты решения?




</div>


## Разбейте данные на выборки

Зададим для данных целевой признак (это тариф Смарт или Ультра), т.е. столбец 'is_ultra' - target, и признаки без целевого - features.

In [4]:
target = df['is_ultra']
features = df.drop(['is_ultra'], axis=1)

**Разобьём данные на обучающие, валидационные и тестовые в соотношении 3:1:1**

Разбиение данных на обучающие (80% от исходных: в обучающие данные сами обучающие данные - 60% и валидационная - 20%) и тестовые (20%), т.е. в соотношении 4:1

In [5]:
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=.2, random_state=12345)

Разбиение обучающих данных на обучающие (75% или 60% от исходных) и валидационные (25% или 20% от исходных) в соотношении 3:1

In [6]:
features_train, features_valid, target_train, target_valid = train_test_split(features_train, target_train, test_size=.25, random_state=12345)

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Все правильно!


<div class="alert alert-warning">



Совет: 


- Обрати внимание на аргумент stratify, он позволит сохранить изначальное распределение таргетов во всех новых датасетах.  Существующий дисбаланс никуда не денется, но в каждом датасете он будет одинаковым. [Почитать](https://pythonru.com/baza-znanij/sklearn-train-test-split) можно тут


- После разбиения лучше перестраховаться и использовать .shape, для контроля за корректностью разбиения


## Исследуйте модели

**Исследуем модель "дерево решений"**

In [7]:
best_model_tree = None
best_result_tree_train = 0 
best_result_tree_valid = 0
best_time_tree = -1
#best_result_tree_test = 0

#Зададим гиперпараметр "максимальная глубина дерева" - depth, который будет меняться в цикле от 1 до 5 включительно
for depth in range(1,6):
    #Зададим модель и обучим её
    time_start = time.time()
    model = DecisionTreeClassifier(random_state=12345, max_depth = depth)
    model.fit(features_train, target_train)
    
    #Найдем предсказания для всех выборок
    predictions_train = model.predict(features_train)
    predictions_valid = model.predict(features_valid)
    #predictions_test = model.predict(features_test)
    
    #Определим качество предсказаний для валидационной выборки с помощью функции accuracy_score
    result_valid = accuracy_score(target_valid, predictions_valid)
    
    #Выберем лучшее из всех
    if result_valid > best_result_tree_valid:
        best_model_tree = model
        best_result_tree_train = accuracy_score(target_train, predictions_train)
        best_result_tree_valid = result_valid
        time_end = time.time()
        best_time_tree = (time_end-time_start)* 10**3
        #best_result_tree_test = accuracy_score(target_test, predictions_test)
        
print('Обучающая выборка:', best_result_tree_train)
print('Валидационная выборка:', best_result_tree_valid)
print(f'Время выполнения: {best_time_tree} ms')
#print('Тестовая выборка:', best_result_tree_test)


Обучающая выборка: 0.8117219917012448
Валидационная выборка: 0.7651632970451011
Время выполнения: 5.757570266723633 ms


<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>



Ошибка ❌:





Пока не выберем лучшую модель на валидации, тестовой выборки для нас как будто бы не существует

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



Исправлено




</div>


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:
 
Работа выполнена в соответствии с критериями: 



 - модель обучена на обучающем наборе
 - получена оценка качества на валидационном наборе
 - перебор гиперпараметров осуществляется в цикле





<div class="alert alert-warning">


Совет: 



- Можно добавить график, хорошо оформленный график украсит проект. Тут на графике можно посмотреть как при изменении гиперпараметров меняется точность на train и validation датасете (для этого собираем наши метрики на валидации и train в списки, а затем по ним обычный plt.plot()) Такой способ используют чтобы лучше отследить переобучение, но для этого стоит взять пернебор по max_depth поглубже



- tree_plot можно построить, с max_depth равной 3 или 4 (глубже уже будет громоздко) - глянуть как DT сплиты делает, будет понимание как модель принимает решения: какие признаки  использует, на каких значениях делает разбиение вправо - влево 




</div>



**Исследуем модель "случайный лес"**

In [8]:
best_model_forest = None
best_result_forest_train = 0
best_result_forest_valid = 0
best_time_forest = -1
#best_result_forest_test = 0

#Зададим гиперпараметр "количество деревьев в лесу" - est, который будет меняться в цикле от 1 до 10 включительно
for est in range(1,15):
    #Зададим модель и обучим её
    time_start = time.time()
    model = RandomForestClassifier(random_state=12345, n_estimators=est)
    model.fit(features_train, target_train)
    
    #Определим качество предсказаний для валидационной выборки с помощью функции model.score()
    result = model.score(features_valid, target_valid)
    
    #Выберем лучшее из всех
    if result > best_result_forest_valid:
        best_model_forest = model
        best_result_forest_train = model.score(features_train, target_train)
        best_result_forest_valid = result
        time_end = time.time()
        best_time_forest = (time_end-time_start)* 10**3
        #best_result_forest_test = model.score(features_test, target_test)
        
print('Обучающая выборка:', best_result_forest_train)
print('Валидационная выборка:', best_result_forest_valid)
print(f'Время выполнения: {best_time_forest} ms')

#print('Тестовая выборка:', best_result_forest_test)
        

Обучающая выборка: 0.9797717842323651
Валидационная выборка: 0.7884914463452566
Время выполнения: 56.32209777832031 ms



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Верно. 

<div class="alert alert-warning">

Совет:


- Иван, можно было сделать похитрее, RF одна из лучших моделей в классическом machine-learning, поэтому можно было добавить перебор 2 гиперпараметров в двойном цикле, в результата метрика качества вырастет. 
 

- Когда что то долго крутиться, можно использовать  %%time - ставишь на самый вверх ячейки с кодом, время выполнения которого хочешь замерить, может не знаешь.  Быстрее не станет, но все будут видеть стоит ли ждать не отходя от ПК или можно сходить чаек поставить )) 
    
    Или tqdm, это ещё лучше, потому что он показывает интерактивно, на каком этапе расчетов мы находимся

    
    from tqdm import tqdm


    for n_estimators in tqdm(range(3,58)):

        ..........

    
И будет красиво )   
 


**Комментарий:** вместо %%time решил использовать библиотеку time c функцией time, т.е. time.time()

*Устар.* Можно заметить, что качество тестовых выборок в обоих моделях одинаковое.

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>



Ошибка ❌:





Забыли о тестовый выборке

**Исследуем модель "логистическая регрессия"**

In [9]:
#import warnings
#warnings.filterwarnings('ignore')
#Поднял import выше

best_model_log = None
best_result_log_train = 0
best_result_log_valid = 0
best_time_log = -1
#best_result_log_test = 0

#Зададим гиперпараметр "количество итераций" - iter, который будет меняться в цикле от 1 до 100 включительно
for iter in range(1,101):
    #Зададим модель и обучим её
    time_start = time.time()
    model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=iter)
    model.fit(features_train, target_train)
    
    #Определим качество предсказаний для валидационной выборки с помощью функции model.score()
    result = model.score(features_valid, target_valid)
    
    #Выберем лучшее из всех
    if result > best_result_log_valid:
        best_model_log = model
        best_result_log_train = model.score(features_train, target_train)
        best_result_log_valid = result
        time_end = time.time()
        best_time_log = (time_end-time_start)* 10**3
        #best_result_log_test = model.score(features_test, target_test)
        
print('Обучающая выборка:', best_result_log_train)
print('Валидационная выборка:', best_result_log_valid)
print(f'Время выполнения: {best_time_log} ms')
#print('Тестовая выборка:', best_result_log_test)

Обучающая выборка: 0.7422199170124482
Валидационная выборка: 0.7371695178849145
Время выполнения: 24.571895599365234 ms


<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>



Ошибка ❌:



Не забываем поднимать импорты наверх

</div>

Исходя из 3 моделей ("дерево решений", "случайный лес", "логистическая регрессия") лучше всего показала 2-я модель, у которой лучше всех качество предсказаний по валидационной выборке, но, помимо этого, у неё также лучшее качество и в тестовой выборке наравне с "деревом решений". Следовательно, модель "случайный лес" более подходящая для предсказаний выбора пользователем тарифа ("Смарт" или "Ультра") на основе предложенных данных. 


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Все верно, RF лучший. 


<div class="alert alert-warning">


Совет: 

Если есть желание можешь ответить на вопросики ))    



- Как назвать ситуацию, когда на валидации при увеличении глубины дерева (модель DT), метрика качества стала снижаться? Как по твоему в это же время вела себя эта же метрика на train?


- Насколько знаю вы еще под капот моделям не заглядывали, но может знаешь почему обычно RF показывает более высокие результаты из выбранных?


- Мы решаем задачу классификации (а еще есть задача регрессии), а в названии модели с помощью которой мы решаем задачу "классификации" (Логистическая регрессия) есть слово "регрессия".  Нет ли тут парадокса? )
    

- Почему логистическая регрессия показывать на много более худшие результаты?  


</div>





**Ответы на вопросы:**
2. По-моему, модель "случайный лес" - смесь "деревьев решений", т.е. эта модель лучше последней
3. Модель логистической регрессии нужна для прогнозирования результатов 0/1, да/нет или правда/ложь, целевой признак "is_ultra" как и является таковым.

В остальном пока затрудняюсь ответить

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>


Успех:


Дополню чуть

1. Переобучение. На train точность растет вплоть до 100%. 
    
    
2. Да, в RF обучается несколько деревьев, а решение принимается путем голосования. Каждое дерево в RF учится на своем датасете (используется техника бутстрап), таким образом достигается "независимость" этих самых деревьев, в итоге ошибка разброса снижается.

    
3. "Регрессия", потому что внутри нее скалярное произведение (как и в линейной регрессии), дающее на выходе вещественное число (от + бесконечности до - бесконечности), на это число вешают сигмойду, логарифм и получают значение от 0 до 1 (вероятность класса).   
    
    
4. Логистическая регрессия плоха (и хороша с другой стороны) тем что это линейный классификатор: обьекты на классы разделяем гиперплоскостью (прямой если на плоскости), а в данном датасете видимо более сложные, нелинейные связи между признаками и таргетом.   А RF с DT с этим справляются. На других данных  LR могла оказаться лучше, например потому что RF не может экстраполировать. 


</div>

## Проверьте модель на тестовой выборке

*Устар.* Данный шаг был сделан в разделе выше, причём на всех моделях.

В предыдущем пункте выяснено, что модель RF является лучшей. Проверим эту модель на тестовой выборке.

In [10]:
#Случайный лес
start = time.time()
result_forest_test = best_model_forest.score(features_test, target_test)
end = time.time()
time_forest = (end-start)*10**3

print('Анализ модели RF на тестовой выборке')
print(f'Случайный лес: качество - {result_forest_test}, время - {time_forest} ms')

Анализ модели RF на тестовой выборке
Случайный лес: качество - 0.7869362363919129, время - 5.665063858032227 ms


Данная модель показала, что она может верно предсказывать выбор клиентом нужного для себя тарифа с вероятностью 78.7%, что является неплохим результатом.



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV3</b></font>

Успех:

Если студент получил на тесте accuraсy  выше 0,78, это считается отличным результатом.





</div>

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Ошибка ❌:




Тестируем только лучшую модель
Это делается для того, чтобы мы даже незначительным образом не "подгонялись" под тестовую выборку. Ведь на train модели обучаются, по валидиации подгоняются гиперпараметры. Эти данные модели "знают". А test (out-of-sample) это уже моделирование прогноза на реальных данных и ситуации когда у нас есть уже лучшая модель (в рельности у нас же не может быть несоклько прогнозов, что то в любом случаи надо выбирать). Вот поэтому такая двухуровневая проверка на подгонку. Кроме того использование мноих моделей с разными гиперпараметрами это тоже подгонка, поэтому выбирая одну и тестируя только ее, мы тем самым боремся с подгонкой через использование многих-многих моделей, когда результат хорош не потому что мы данные почистили хорошо, моделировали правильно итд итп, а потому что из многих моделей хоть какая то случайно "сыграет".

Видим, что по качеству одинаково хорошо показали себя две модели: "дерево решений", "случайный лес", но и последняя третья модель ("логистическая регрессия") тоже далеко не ушла, она чуть похуже; по скорости лучше всего вышла третья модель, чуть помедленнее сработало "дерево решений", в роли "черепахи" выступила модель "случайный лес", но эта разница во времени не заметна для человеческого глаза, учитывая, что 1 ms ("миллисекунда") = 0.001 s ("секунда") 


<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>




Ошибка:


А должны сделать это тут

Иван, о логике использования датасетов:
    




1. На train мы обучаем
2. По валидации смотрим на результаты обучения (следим чтобы не было переобучения и/или делаем подбор гиперпараметров).  И выбираем лучшую модель. 
3. Тестовая (out-of_sample) у нас для финальной проверки, когда определена лучшая по валидации модель с конкретными гиперпараметрами. 
    
    
    
 [Вот](https://towardsdatascience.com/why-do-we-need-a-validation-set-in-addition-to-training-and-test-sets-5cf4a65550e0   ) тут можно дополнительно почитать.


## (бонус) Проверьте модели на адекватность

In [11]:


f1_score_tree = f1_score(target_valid, best_model_tree.predict(features_valid))
f1_score_forest = f1_score(target_valid, best_model_forest.predict(features_valid))
f1_score_log = f1_score(target_valid, best_model_log.predict(features_valid))

print(f'Дерево решений: accuracy - {best_result_tree_valid}, f1_score - {f1_score_tree}')
print(f'Случайный лес: accuracy - {best_result_forest_valid}, f1_score - {f1_score_forest}')
print(F'Логистическая регрессия: accuracy - {best_result_log_valid}, f1_score - {f1_score_log}')

Дерево решений: accuracy - 0.7651632970451011, f1_score - 0.5113268608414241
Случайный лес: accuracy - 0.7884914463452566, f1_score - 0.6023391812865497
Логистическая регрессия: accuracy - 0.7371695178849145, f1_score - 0.32128514056224905


<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Ошибка ❌:





Импорт поднимаем наверх

Исходя из этого, можно считать, что модель "случайный лес" лучше показала себя на адекватность, т.к. у неё больше всех значение f1_score. 


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Совет:


Иван, тобой решено проверить модель на адекватность "подгрузив" несколько других метрик качества. Да, в этом резон есть, например нам важно было качество предсказания какого-то одного класса, тогда да, мы бы проверяли на метриках, которые учитывают это и получили бы более адекватные оценки.   

Но в данном случаи мы проверяем модель на адекватность сравнивая ее с наивным прогнозом (простой "угадалкой" или прогноз одного и того же класса). Не забудь у нас дисбаланс классов в таргете. 


При проверке на адекватность модели, рекомендую ознакомиться с [Dummy Models](https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html), это удобная обертка для константных моделей


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Совет:


Тут логика такая - у нас есть данные и мы видим что доля нулей в таргете около 69%, это значит что мы не проводя никакого анализа, моделирования итп итд, можем постоянно предсказывать 0 и получить точность около 69%. Вот эту accuracy в 69% мы и должны побить, иначе окажется что наша модель не лучше наивного прогноза самого частого класса.
      
    
С   DummyClassifier это выглядит  так:
    
    DummyClassifier(strategy='most_frequent', random_state=1)  
задаем константную модель, которая будет предсказывать по most_frequent - самый частый класс
    
    .fit(features_train, target_train)
тут модель смотри на самый частый класс в  target_train - типо "обучение"
    
    .predict(features_test)
    .score(features_test, target_test)
проогнозирует везде самый частый класс в features_test и считает score
    
Получаем тот же результат: acc около 69%.    



Сравниваем с accuracy нашей модели и делаем вывод


## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75 *(за исключением модели "логистическая регрессия" - приближенно 0.74)*


## Общий вывод

**Итоговый вывод:** В данном проекте был изучены данные клиентов оператора мобильной связи "Мегалайн", где выяснилось, что не требуется предобработка данных.

Данные были разбиты по столбцам на целевой признак (ответы) - столбец 'is_ultra' и признаки(вопросы) - все остальные столбцы, по строкам на обучающие, валидационные и тестовые данные в соотношении 3:1:1 (= 60%:20%:20%)

Далее были исследованы три модели машинного обучения, связанные с классификацией данных: "дерево решений", "случайный лес", "логистическая регрессия". Лучше всего из 3 моделей показала результат на качество и вменяемость - модель "случайный лес".

Таким образом, оператору мобильной связи "Мегалайн" следует выбрать модель предсказаний "случайный лес" для построения системы, способной проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Иван, здорово что в конце расписаны все этапы работы. Это важно потому что когда проект захочет посмотреть будущий работодатель (или начальник), у него может не быть времени на подробный разбор кода. Вероятнее всего он бегло просмотрит код, а из общего вывода захочет получить представление о всей работе.



</div>



<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>




Иван, у тебя старательно выполненная работа, все четко, осмысленно. Выводы присутствуют, с комментированием кода тоже никаких проблем нет, твоих объяснений достаточно для понимания коллегами хода твоих мыслей. 




Я оставил небольшие советы и вопросики (если есть время и желание можешь воспользоваться/ответить): 




- обрати внимание на проверку сбалансированность классов в таргете в разделе EDA
- использовать stratify
- добавить график метрики    
- как можно улучшить результаты 



    
Обязательное к исправлению:




- импорты собираем в одном месте, так будет удобней и тебе и тем коллегам кто будет работать с твоим кодом    


    
- стоит переделать бонусное задание с учётом моего комментария (по желанию, это все таки бонусное)
 
    
    
- нарушена логика использования датасетов   -  на тестовой выборке тестируем только лучшую модель 

    
    
Жду исправлений, для принятия проекта. Если какие то вопросы, то сразу спрашивай ) 





</div>


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Спасибо за работу!    

    


Что осталось из красного:

 

    
- Финальное тестирование проводим только на лучшей модели
    
    
- Импорты поднимаем наверх    


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV3</b></font>

Спасибо за работу!    

    
Красное исправлено.
  
Отличная работа Иван. Желаю успехов в дальнейшей учебе!
